### ISSUE TITLE
Lighter weight Request Options

### ISSUE THREAD

**Issue by t8y8** [tableau/server-client-python#114]
Today in the code we have request options for various requests, some applicable to many endpoints, some only one...

- pageSize
- pageNumber
- filters
- sort order
- flags (overwrite=T, resolution=high)

They're implemented, in most cases, as nested enums in `request_options.py` requiring sequences like this in user-code:
```
req_option.filter.add(TSC.Filter(TSC.RequestOptions.Field.Name,
                                   TSC.RequestOptions.Operator.Equals, args.view_name))
```
```
request_options = TSC.RequestOptions(pagesize=1000)
all_workbooks = list(TSC.Pager(server.workbooks, request_options))
```

I'm opening this issue to track discussion of other options.

**Comment by t8y8** [comment]
I went ahead and sketched out several ideas and examples for comparison and discussion. 
I'm not sure that I've got a strong preference on any yet.

```python
##################
###  Option 0 -- Today

# Name equals 'tyler' with a pageSize of 1000
request_options = TSC.RequestOptions(pagesize=1000)
request_options.filter.add(TSC.Filter(TSC.RequestOptions.Field.Name, TSC.RequestOptions.Operator.Equals, 'tyler')
all_workbooks = list(TSC.Pager(server.workbooks, request_options))

# lastLogin is greater than or equal to 12/13/1988 with a page size of 10 starting on page 2
request_options = TSC.RequestOptions(pagesize=10, pageNumber=2)
request_options.filter.add(TSC.Filter(TSC.RequestOptions.Field.LastLogin, TSC.RequestOptions.Operator.GreaterThanOrEqual, "1988-12-13 00:00:00")
all_workbooks = list(TSC.Pager(server.workbooks, request_options))


####################
### Option 1 -- Format Strings

# Name equals 'tyler' with a pageSize of 1000
request_options = TSC.RequestOptions("pageSize=1000", "name:eq:tyler")
all_workbooks = list(TSC.Pager(server.workbooks, request_options))

# lastLogin is greater than or equal to 12/13/1988 with a page size of 10 starting on page 2
request_options = TSC.RequestOptions("pageSize=10", "pageNumber=2", "lastLogin:gte:1988-12-13 00:00:00")
all_workbooks = list(TSC.Pager(server.workbooks, request_options))

########################
### Option 2 -- Validating Dictionary

# Name equals 'tyler' with a pageSize of 1000
request_options = TSC.RequestOptions({"pageSize": 1000, "name": "tyler"}
all_workbooks = list(TSC.Pager(server.workbooks, request_options))

# lastLogin is greater than or equal to 12/13/1988 with a page size of 10 starting on page 2
request_options = TSC.RequestOptions({"pageSize": 10, "pageNumber": 2, "lastLogin": ">=1988-12-13 00:00:00")
all_workbooks = list(TSC.Pager(server.workbooks, request_options))

################
### Option 3 -- Fluid API
# Name equals 'tyler' with a pageSize of 1000
request_options = TSC.RequestOptions.pageSize(1000).Name('=', 'tyler')
all_workbooks = list(TSC.Pager(server.workbooks, request_options))

# lastLogin is greater than or equal to 12/13/1988 with a page size of 10 starting on page 2
request_options = TSC.RequestOptions.pageSize(10).pageNumber(2).LastLogin('>=', "1988-12-13 00:00:00")
all_workbooks = list(TSC.Pager(server.workbooks, request_options))
```

**Comment by RussTheAerialist** [comment]
Fwiw, I'm a fan of the fluid API. It's clean, easy to read, mostly devoid
of extra cruft.

The problem that I have with the dictionary approach is that it provides no
help to the end user about what is available, and simple typos in the key
name could lead to confusing behavior (unless we explicitly reject unknown
keys, but at that point we'd be reimplementing something that using method
calls would just handle.)

**Comment by t8y8** [comment]
Here's a basic prototype that implements option 3.

But it requires more meta-programming than I currently have a handle on to understand to get rid of the ()'s:
```python
class Fluent2:
    def __init__(self, cache=None):
        self._expressions = []

    # Final method call
    def build_url(self):
        print(self._expressions)
        return '&'.join(self._expressions)

    def Name(self, name=None):
        self._expressions.append('name:')
        return self

    def equals(self, value):
        self._expressions[-1] += ":".join(['eq', value])
        return self

    def pageSize(self, value):
        self._expressions.append("pageSize={}".format(value))
        return self

options = Fluent2().Name().equals('tyler').pageSize(1337)
print(options.build_url())
```

**Comment by RussTheAerialist** [comment]
Can we wait on this until I get back to the office?

**Comment by t8y8** [comment]
Ha, yeah we can :)

(I'm digging in because I need something to do over the holidays but I'm in no rush to implement this).

Here' a slightly improved version:
```python
class RequestOptions2:

    def __init__(self):
        self._expressions = []
    
    @property
    def Name(self):
        self._expressions.append('name:')
        return self

    @property
    def LastLogin(self):
        self._expressions.append('lastLogin:')
        return self

    def equals(self, value):
        self._expressions[-1] += ':'.join(['eq', str(value)])
        return self

    def gte(self, value):
        self._expressions[-1] += ':'.join(['gte', str(value)])
        return self

    def __str__(self):
        # This could do validation or someone else could
        return "&".join(self._expressions)


blah = RequestOptions2().Name.equals('tyler').LastLogin.gte('1988-12-13')
print(blah)
```

**Comment by t8y8** [comment]
I've implemented several versions of this that works with some level of validation. 
Just posting here so I don't forget. I make no claims that running this code won't summon cthulhu

```python
class RequestOptions3:

    def __init__(self, **kwargs):
        self._expressions = {}
        self.parameters = {}
        for k, v in kwargs.items():
            self.parameters[k] = v
        self._in_progres = None
    
    @property
    def Name(self):
        self._expressions['name'] = ()
        self._in_progres = 'name'
        return self

    @property
    def LastLogin(self):
        self._expressions['lastLogin'] = ()
        self._in_progres = 'lastLogin'
        return self

    @property
    def CreatedAt(self):
        self._expressions['createdAt'] = ()
        self._in_progres = 'createdAt'
        return self

    @property
    def LastLogin(self):
        self._expressions['lastLogin'] = ()
        self._in_progres = 'lastLogin'
        return self

    @property
    def OwnerName(self):
        self._expressions['ownerName'] = ()
        self._in_progres = 'ownerName'
        return self

    @property
    def SiteRole(self):
        self._expressions['siteRole'] = ()
        self._in_progres = 'siteRole'
        return self

    @property
    def Tags(self):
        self._expressions['tags'] = ()
        self._in_progres = 'tags'
        return self

    @property
    def UpdatedAt(self):
        self._expressions['updatedAt'] = ()
        self._in_progres = 'updatedAt'
        return self


    def eq(self, value):
        self._expressions[self._in_progres] = ('eq', str(value))
        return self

    def gte(self, value):
        self._expressions[self._in_progres] = ('gte', str(value))
        return self

    def lte(self, value):
        self._expressions[self._in_progres] = ('lte', str(value))
        return self

    def lt(self, value):
        self._expressions[self._in_progres] = ('lt', str(value))
        return self

    def gt(self, value):
        self._expressions[self._in_progres] = ('gt', str(value))
        return self

    def In(self, value):
        values = "[{}]".format(','.join(str(i) for i in value))
        self._expressions[self._in_progres] = ('in', str(values))
        return self

    def _validate(self):
        for field, (expression, value) in self._expressions.items():
            if not expression:
                raise Exception
            if expression == "in" and not value.count('[') >= 1:
                # Crappy test for is in looking at a list
                raise Exception

    def _make_filters(self):
        filters = []
        for k, v in self._expressions.items():
            f = ":".join([k, v[0], v[1]])
            filters.append(f)
        
        return "filters=" + ",".join(filters)

    def __str__(self):
        filters = self._make_filters()
        paramerers = "&".join("{}={}".format(k, v) for k, v in self.parameters.items())
        return "&".join([paramerers, filters])

blah = RequestOptions3(pageSize=100, pageNumber=1000, a='b').Name.eq('tyler').In([1,2,3]).LastLogin.gte('1988-12-13')
print(blah._validate())
print(blah)

_DUMMY = object()

class RequestOptions4:

    def __init__(self, **kwargs):
        self._expressions = {}
        self.parameters = {}
        for k, v in kwargs.items():
            self.parameters[k] = v
    
    @property
    def Name(self):
        self._expressions['name'] = _DUMMY
        return self

    @property
    def LastLogin(self):
        self._expressions['lastLogin'] = _DUMMY
        return self

    @property
    def CreatedAt(self):
        self._expressions['createdAt'] = _DUMMY
        return self

    @property
    def LastLogin(self):
        self._expressions['lastLogin'] = _DUMMY
        return self

    @property
    def OwnerName(self):
        self._expressions['ownerName'] = _DUMMY
        return self

    @property
    def SiteRole(self):
        self._expressions['siteRole'] = _DUMMY
        return self

    @property
    def Tags(self):
        self._expressions['tags'] = _DUMMY
        return self

    @property
    def UpdatedAt(self):
        self._expressions['updatedAt'] = _DUMMY
        return self


    def eq(self, value):
        key = next((field for field, value in self._expressions.items() if value == _DUMMY), None)
        self._expressions[key] = ('eq', str(value))
        return self

    def gte(self, value):
        key = next((field for field, value in self._expressions.items() if value == _DUMMY), None)
        self._expressions[key] = ('gte', str(value))
        return self

    def lte(self, value):
        key = next((field for field, value in self._expressions.items() if value == _DUMMY), None)
        self._expressions[key] = ('lte', str(value))
        return self

    def lt(self, value):
        key = next((field for field, value in self._expressions.items() if value == _DUMMY), None)
        self._expressions[key] = ('lt', str(value))
        return self

    def gt(self, value):
        key = next((field for field, value in self._expressions.items() if value == _DUMMY), None)
        self._expressions[key] = ('gt', str(value))
        return self

    def In(self, value):
        key = next((field for field, value in self._expressions.items() if value == _DUMMY), None)
        values = "[{}]".format(','.join(str(i) for i in value))
        self._expressions[key] = ('in', str(values))
        return self

    def _validate(self):
        for field, (expression, value) in self._expressions.items():
            if expression is None or value is None or field is None:
                import pprint
                raise Exception("Invalid expression tree. \nLook for `None` for the problmatic expression\n" +
                                pprint.pformat(self._expressions))

    def _make_filters(self):
        filters = []
        for f, (e, v) in self._expressions.items():
            f = ":".join([f, e, v])
            filters.append(f)
        
        return "filters=" + ",".join(filters)

    def __str__(self):
        self._validate()
        filters = self._make_filters()
        paramerers = "&".join("{}={}".format(k, v) for k, v in self.parameters.items())
        return "&".join([paramerers, filters])


blah = RequestOptions4(pageSize=100, pageNumber=1000, a='b').Name.eq('tyler').Tags.In([1,2,3]).LastLogin.gte('1988-12-13')
print(blah._validate())
print(blah)
```

**Comment by mnraj18** [comment]
Hi ,

I am just exploring TSC and would like to get some help on the issue "Generator Expression Must be Parenthesized" while trying to list out workbooks updated sometime ago.. My code is as below ..
------------------------------------------
request_options=TSC.RequestOptions()
request_options.filter.add(TSC.Filter(TSC.RequestOptions.Field.UpdatedAt, 
                                      TSC.RequestOptions.Operator.GreaterThanOrEqual, 
                                      #01-Dec-2020#)
all_workbooks=list(TSC.Pager(server.workbooks, request_options))
print(wb.name for wb in all_workbooks)
-----------------------------------------
Also, may i know what are the print options we have to print list/list of items, awaiting your revert please ?

**Comment by bcantoni** [comment]
@mnraj18 could you please write that as a new issue rather than a comment here?

**Comment by mnraj18** [comment]
Yes, Sorry..
